AEPLifecycle/Sources/XDM/XDMLifecycleMetricsBuilder.swift

```diff
@@ -0,0 +1,137 @@
+    /// - Returns: App close event data in dictionary format
+    func buildXDMAppCloseEventData() -> [String: Any]? {
+        var appLaunchXDMData = XDMMobileLifecycleDetails()
+        appLaunchXDMData.application = xdmApplicationInfoClose
+        appLaunchXDMData.eventType = LifecycleConstants.XDM.EVENT_TYPE_APP_CLOSE
+        appLaunchXDMData.timestamp = startDate
+
+        // TODO: MOB-14370 backdate timestamp
+
+        return appLaunchXDMData.asDictionary()
+    }
+
+    /// Adds general application information as well as details related to the type of launch (install, upgrade, regular launch)
+    /// - Parameters:
+    ///   - isInstall: indicates if this is an app install
+    ///   - isUpgrade: indicates if this is an app upgrade
+    /// - Returns: this `XDMLifecycleMetricsBuilder` instance
+    func addAppLaunchData(isInstall: Bool, isUpgrade: Bool) -> XDMLifecycleMetricsBuilder {
+        xdmApplicationInfoLaunch = XDMLifecycleApplication()
+        xdmApplicationInfoLaunch?.isLaunch = true
+
+        if isInstall {
+            xdmApplicationInfoLaunch?.isInstall = true
+            xdmApplicationInfoLaunch?.installDate = startDate
+        }
+
+        if isUpgrade {
+            xdmApplicationInfoLaunch?.isUpgrade = true
+        }
+
+        xdmApplicationInfoLaunch?.name = systemInfoService.getApplicationName()
+        xdmApplicationInfoLaunch?.id = systemInfoService.getApplicationBundleId()
+        xdmApplicationInfoLaunch?.version = systemInfoService.getApplicationVersion()
+
+        return self
+    }
+
+    /// Adds general application information as well as details related to the type of close event
+    /// - Parameters:
+    ///   - previousAppId: application version from the previous session when the close happened
+    ///   - previousSessionInfo: previous session information to be attached to the close event
+    /// - Returns: this `XDMLifecycleMetricsBuilder` instance
+    func addAppCloseData(previousAppId: String?, previousSessionInfo: LifecycleSessionInfo) -> XDMLifecycleMetricsBuilder {
+        xdmApplicationInfoClose = XDMLifecycleApplication()
+
+        xdmApplicationInfoClose?.version = previousAppId
+        xdmApplicationInfoClose?.isClose = true
+        xdmApplicationInfoClose?.closeType = previousSessionInfo.isCrash ? .unknown : .close
+        if let sessionLength = previousSessionInfo.sessionLength {
+            xdmApplicationInfoClose?.sessionLength = Int64(sessionLength)
+        }
+        xdmApplicationInfoClose?.name = systemInfoService.getApplicationName()
+        xdmApplicationInfoClose?.id = systemInfoService.getApplicationBundleId()
+
+        return self
+    }
+
+    /// Adds information related to the running environment, see `XDMLifecycleEnvironment`
+    /// - Returns: this `XDMLifecycleMetricsBuilder` instance
+    func addEnvironmentData() -> XDMLifecycleMetricsBuilder {
+        xdmEnvironmentInfo = XDMLifecycleEnvironment()
```

Comment:
for all add* functions it looks like we reinstantiate. Is this your intention? Should we document this behavior or even reject subsequent calls to the same function in this builder?

Reply:
Yeah, this is intentional, I think it makes sense as the object it creates in the function is only ever modified from within that function.